File: letterCombinationsOfAPhoneNumber/letterCombinations_dfs.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<string> letterCombinations(string digits) {
        vector<string> res;
        if (digits.length() == 0)
            return res;
        vector<string> vec = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        string path;
        dfs(digits, res, path, vec, 0);
        return res;
    }
    void dfs(string& digits, vector<string>& res, string& path, vector<string>& vec, int index) {
        if (index == digits.size()) {
            res.push_back(path);
            return;
        }
        for (char c : vec[digits[index] - '0']) {
            path.push_back(c);
            dfs(digits, res, path, vec, index + 1);
            path.pop_back();
        }
    }
};

int main()
{
    Solution s;
    vector<string> ans = s.letterCombinations(string("234"));
    vector<string>::iterator it = ans.begin();
    while(it != ans.end()){
        cout << *it << " ";
        ++it;
    }
    cout << endl;
    return 0;
}

File: letterCombinationsOfAPhoneNumber/letterCombinations_dfs.java
class Solution {
    public List<String> letterCombinations(String digits) {
        List<String> res = new ArrayList<String>();
        if (digits == null || digits.length() == 0) {
            return res;
        }
        String[] digit2letters = new String[] {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        StringBuilder path = new StringBuilder();
        dfs(digits, path, 0, res, digit2letters);
        return res;
    }
    void dfs(String digits, StringBuilder path, int level, List<String> res, String[] digit2letters) {
        if (level == digits.length()) {
            res.add(path.toString());
            return;
        }
        String letters = digit2letters[digits.charAt(level) - '0'];
        if (letters.length() == 0)
            dfs(digits, path, level + 1, res, digit2letters);
        for (int i = 0; i < letters.length(); i++) {
            char c = letters.charAt(i);
            path.append(c);
            dfs(digits, path, level + 1, res, digit2letters);
            path.deleteCharAt(path.length() - 1);
        }
    }
}
